test-data/unit/check-newsemanal.test

```diff
@@ -1856,3 +1858,47 @@ reveal_type(y[0])  # E: Revealed type is 'builtins.int'
 x: A
 reveal_type(x)  # E: Revealed type is '__main__.G[Tuple[builtins.int, fallback=__main__.C]]'
 [builtins fixtures/list.pyi]
+
+[case testNewAnalyzerCastForward]
+from typing import cast
+
+x = cast('C', None)
+class A:
+    def foo(self) -> None:
+        self.x = cast('C', None)
+
+reveal_type(x)  # E: Revealed type is '__main__.C'
+reveal_type(A().x)  # E: Revealed type is '__main__.C'
+
+class C(A): ...
```

Comment:
Maybe add couple of tests that use special forms (type aliases, named tuples, etc)?